torch/testing/_internal/common_modules.py

```diff
@@ -0,0 +1,203 @@
+                    raise ex
+
+
+def formatted_module_name(module_cls):
+    """ Returns the common name of the module class formatted for use in test names. """
+    return MODULE_CLASS_NAMES[module_cls].replace('.', '_')
+
+
+class FunctionInput(object):
+    """ Contains args and kwargs to pass as input to a function. """
+    def __init__(self, *args, **kwargs):
+        self.args = args
+        self.kwargs = kwargs
+
+
+class ModuleInput(object):
+    """ Contains args / kwargs for module instantiation + forward pass. """
+    __slots__ = ['constructor_input', 'forward_input', 'desc', 'reference_fn']
+
+    def __init__(self, constructor_input, forward_input=None, desc='', reference_fn=None):
+        self.constructor_input = constructor_input  # Inputs to pass during construction
+        self.forward_input = forward_input  # Inputs to pass to forward()
+        self.desc = desc  # Description for this set of inputs
+        self.reference_fn = reference_fn  # Reference function for comparison; expected signature:
+                                          # reference_fn(module, parameters, *args, **kwargs)
+
+        if reference_fn is not None:
+
+            @wraps(reference_fn)
+            def copy_reference_fn(m, *args, **kwargs):
+                # Copy inputs to avoid undesired side effects from calling the reference.
+                args, kwargs = deepcopy(args), deepcopy(kwargs)
+
+                # Note that module parameters are passed in for convenience.
+                return reference_fn(m, list(m.parameters()), *args, **kwargs)
+
+            self.reference_fn = copy_reference_fn
+
+
+class ModuleInfo(object):
+    """ Module information to be used in testing. """
+
+    def __init__(self,
+                 module_cls,  # Class object for the module under test
+                 *,
+                 module_inputs_func,  # Function to generate module inputs
+                 skips=[],  # Indicates which tests to skip
+                 decorators=None,  # Additional decorators to apply to generated tests
+                 ):
+        self.module_cls = module_cls
+        self.module_inputs_func = module_inputs_func
+        self.skips = skips
+        self.decorators = decorators
+
+    def should_skip(self, cls_name, test_name, device_type, dtype):
+        return any(si.is_active(cls_name, test_name, device_type, dtype) for si in self.skips)
+
+    @property
+    def name(self):
+        return formatted_module_name(self.module_cls)
+
+
+def module_inputs_torch_nn_Linear(module_info, device, dtype, requires_grad, **kwargs):
+    make_input = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+
+    module_inputs = [
+        ModuleInput(constructor_input=FunctionInput(10, 8),
+                    forward_input=FunctionInput(make_input((4, 10))),
+                    reference_fn=lambda m, p, i: torch.mm(i, p[0].t()) + p[1].view(1, -1).expand(4, 8)),
```

Comment:
Interesting. In the future I wonder if the reference can just a functional form of the module so you don't need to write a reference for each input?

Reply:
The ability to write a reference per input is actually a feature. For a given (constructor args, forward args), the reference function may be as simple or complex as needed to verify the output. I think it'd be non-ideal to require that the module is essentially reimplemented within the reference, and if we just call the existing functional form of the module, that doesn't really test much. I'd also like to use the references from `module_tests` / `new_module_tests` / `criterion_tests` as much as possible.

Reply:
Sure, pros/cons, and probably not too hard to change